### PR TITLE
V6 Optimizations

### PR DESCRIPTION
**Pull request by jayanthvn**
*Issue #, if available:* N/A

*Description of changes:*
With the [latest AMI](https://github.com/awslabs/amazon-eks-ami/releases/tag/v20230919) V6 tc programs were failing to load hence optimizing some of the v6 code.

```
uname -r
5.10.192-183.736.amzn2.x86_64
```

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: pkg/ebpf/c/tc.v6egress.bpf.c [hunk 1]
```diff
@@ -119,119 +115,109 @@ int handle_egress(struct __sk_buff *skb)
...
+	//Check for the an existing flow in the conntrack table
+	flow_key.saddr = ip->saddr;
+	flow_key.daddr = ip->daddr;
+	flow_key.src_port = l4_src_port;
+	flow_key.dest_port = l4_dst_port;
+	flow_key.protocol = ip->nexthdr;
+
+	//Check if it's an existing flow
+	flow_val = (struct conntrack_value *)bpf_map_lookup_elem(&aws_conntrack_map, &flow_key);
+	if (flow_val != NULL && (flow_val->addr.in6_u.u6_addr8[12] == flow_key.saddr.in6_u.u6_addr8[12]
```
Comment (dims) [inline pkg/ebpf/c/tc.v6egress.bpf.c:159]: can we use `__builtin_memcmp`?

File: pkg/ebpf/c/tc.v6egress.bpf.c [hunk 2]
```diff
@@ -119,119 +115,109 @@ int handle_egress(struct __sk_buff *skb)
...
-				evt.verdict = 1;
-				bpf_ringbuf_output(&policy_events, &evt, sizeof(evt), 0);
-				return BPF_OK;
-			}
+		if ((trie_val->protocol == ANY_IP_PROTOCOL) || (trie_val->protocol == ip->nexthdr &&
+					((trie_val->start_port == ANY_PORT) || (l4_dst_port == trie_val->start_port) ||
+					(l4_dst_port > trie_val->start_port && l4_dst_port <= trie_val->end_port)))) {
+			//Inject in to conntrack map
+			struct conntrack_value new_flow_val;
+			__builtin_memset(&new_flow_val, 0, sizeof(new_flow_val));
```
Comment (dims) [inline pkg/ebpf/c/tc.v6egress.bpf.c:204]: is there a `__builtin_bzero`

File: pkg/ebpf/c/tc.v6egress.bpf.c [hunk 3]
```diff
@@ -119,119 +115,109 @@ int handle_egress(struct __sk_buff *skb)
...
+	evt.protocol = flow_key.protocol;
+
+	//Check for the reverse flow entry in the conntrack table
+	flow_key.daddr = ip->saddr;
+	flow_key.saddr = ip->daddr;
+	flow_key.src_port = flow_key.dest_port;
+	flow_key.dest_port = l4_src_port;
+		
+	reverse_flow_val = (struct conntrack_value *)bpf_map_lookup_elem(&aws_conntrack_map, &flow_key);
+	if (reverse_flow_val != NULL && (reverse_flow_val->addr.in6_u.u6_addr8[12] == flow_key.daddr.in6_u.u6_addr8[12]
```
Comment (dims) [inline pkg/ebpf/c/tc.v6egress.bpf.c:179]: same as above __builtin_memcmp?

File: pkg/utils/utils.go [hunk 4]
```diff
@@ -296,7 +297,7 @@ type ConntrackVal struct {
 
 func ConvConntrackV6ToByte(key ConntrackKeyV6) []byte {
 	ipSize := unsafe.Sizeof(key)
-	byteArray := (*[38]byte)(unsafe.Pointer(&key))
+	byteArray := (*[40]byte)(unsafe.Pointer(&key))
```
Comment (dims) [inline pkg/utils/utils.go:300]: could we please doc how we got to 40? :)

Comment (dims) [reply pkg/utils/utils.go:300]: Also do we need protection on length of stuff we copy in line 308 as well?

File: pkg/ebpf/c/tc.v6egress.bpf.c (hunk 2 above)
Comment (jayanthvn) [reply pkg/ebpf/c/tc.v6egress.bpf.c:204]: I was reading in one the docs that bzero is getting deprecated - https://www.heatware.net/linux-tips/bzero-solution-incompatible-implicit-declaration/#

File: pkg/ebpf/c/tc.v6egress.bpf.c (hunk 1 above)
Comment (jayanthvn) [reply pkg/ebpf/c/tc.v6egress.bpf.c:159]: Yes I did try using this - 

```
#define IN6_ARE_ADDR_EQUAL(a, b)			\
    (__builtin_memcmp(&(a)->in6_u.u6_addr8[0], &(b)->in6_u.u6_addr8[0], sizeof(struct in6_addr)) == 0)
```

But the compiler wasn't happy hence have kept the #define in the code which we can revisit.

File: pkg/utils/utils.go (hunk 4 above)
Comment (jayanthvn) [reply pkg/utils/utils.go:300]: actually, it is the size of the v6 contract key..I will update it.

File: pkg/ebpf/c/tc.v6egress.bpf.c (hunk 2 above)
Comment (dims) [reply pkg/ebpf/c/tc.v6egress.bpf.c:204]: gotcha!

File: pkg/ebpf/c/tc.v6egress.bpf.c (hunk 1 above)
Comment (dims) [reply pkg/ebpf/c/tc.v6egress.bpf.c:159]: thanks!

**Review by achevuru — APPROVED** [review]

**Pull request merged**